beis_indicators/utils/nuts_utils.py

```diff
@@ -0,0 +1,107 @@
+import json
+import pandas as pd
+import numpy as np
+
+from beis_indicators.utils.set_utils import set_containment
+
+NUTS_ENFORCED = {
+    2003: 2003,
+    2006: 2008,
+    2010: 2012,
+    2013: 2015,
+    2016: 2018
+}
+
+NUTS_IDS ={
+        2003: {'UKC2', 'UKK4', 'UKL2', 'UKG3', 'UKK2', 'UKD1', 'UKF2', 'UKG1', 
+            'UKD3', 'UKH3', 'UKI1', 'UKJ4', 'UKM4', 'UKE2', 'UKI2', 'UKM1', 
+            'UKJ1', 'UKJ3', 'UKM2', 'UKE4', 'UKE1', 'UKK1', 'UKE3', 'UKK3', 
+            'UKG2', 'UKD2', 'UKL1', 'UKJ2', 'UKN0', 'UKH1', 'UKM3', 'UKH2', 
+            'UKF3', 'UKD5', 'UKD4', 'UKC1', 'UKF1'}, 
+        2006: {'UKC2', 'UKK4', 'UKL2', 'UKG3', 'UKK2', 'UKD1', 'UKF2', 'UKG1', 
+            'UKD3', 'UKH3', 'UKI1', 'UKJ4', 'UKE2', 'UKI2', 'UKJ1', 'UKE4', 
+            'UKM2', 'UKJ3', 'UKE1', 'UKK1', 'UKE3', 'UKK3', 'UKG2', 'UKM5', 
+            'UKD2', 'UKL1', 'UKJ2', 'UKN0', 'UKH1', 'UKM3', 'UKF3', 'UKH2', 
+            'UKD5', 'UKM6', 'UKD4', 'UKC1', 'UKF1'}, 
+        2010: {'UKC2', 'UKK4', 'UKL2', 'UKG3', 'UKK2', 'UKD1', 'UKF2', 'UKG1', 
+            'UKD3', 'UKH3', 'UKI1', 'UKJ4', 'UKE2', 'UKI2', 'UKJ1', 'UKE4', 
+            'UKJ3', 'UKM2', 'UKE1', 'UKK1', 'UKE3', 'UKK3', 'UKG2', 'UKM5', 
+            'UKL1', 'UKD7', 'UKJ2', 'UKN0', 'UKH1', 'UKM3', 'UKF3', 'UKH2', 
+            'UKD4', 'UKM6', 'UKC1', 'UKD6', 'UKF1'}, 
+        2013: {'UKC2', 'UKK4', 'UKL2', 'UKG3', 'UKK2', 'UKI6', 'UKD1', 'UKF2', 
+            'UKI7', 'UKG1', 'UKD3', 'UKH3', 'UKJ4', 'UKE2', 'UKJ1', 'UKJ3', 
+            'UKM2', 'UKE4', 'UKE1', 'UKK1', 'UKE3', 'UKI5', 'UKK3', 'UKG2', 
+            'UKM5', 'UKL1', 'UKI4', 'UKD7', 'UKJ2', 'UKN0', 'UKH1', 'UKM3', 
+            'UKH2', 'UKF3', 'UKD4', 'UKM6', 'UKI3', 'UKC1', 'UKD6', 'UKF1'}, 
+        2016: {'UKM7', 'UKC2', 'UKM8', 'UKK4', 'UKL2', 'UKG3', 'UKK2', 'UKI6', 
+            'UKD1', 'UKF2', 'UKI7', 'UKG1', 'UKD3', 'UKH3', 'UKJ4', 'UKE2', 
+            'UKJ1', 'UKJ3', 'UKE4', 'UKE1', 'UKK1', 'UKE3', 'UKI5', 'UKM9', 
+            'UKK3', 'UKG2', 'UKM5', 'UKL1', 'UKI4', 'UKD7', 'UKJ2', 'UKN0', 
+            'UKH1', 'UKF3', 'UKH2', 'UKM6', 'UKD4', 'UKC1', 'UKI3', 'UKD6', 'UKF1'}}
+
+NUTS_YEARS = np.array(list(NUTS_IDS.keys()))
+
+def _year_containments(ids, years):
+    containments = []
+    for year in years:
+        year_ids = NUTS_IDS[year]
+        containments.append(set_containment(ids, year_ids))
+    containments = np.array(containments)
+    return containments
+
+def nuts_earliest(year):
+    '''nuts_earliest
+    Returns the earliest possible NUTS version for a year
+    based on the enforcement date.
+    '''
+    for k, v in NUTS_ENFORCED.items():
+        if year >= v:
+            earliest = k
+    return earliest
+
+def _detect_nuts(ids, year):
```

Comment:
here I'd use `_detect_nuts2_uk`

Reply:
Ok. I think long term these will become generic functions though

Reply:
Long term, tight now they aren't, so we should use naming that is appropriate for this version of the source code.